### PR TITLE
Add concept + move Constructor to @dolittle/types

### PR DESCRIPTION
**Pull request by woksin**
Adds the ConceptAs class and other things needed to create that class.

### CODE REVIEW

File: Source/concepts/Concept.ts [hunk 1]
```diff
@@ -0,0 +1,123 @@
+// Copyright (c) Dolittle. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+import { IEquatableTo, IEquatable } from '@dolittle/rudiments';
+import { typeGuard } from '@dolittle/types';
+
+type ConceptBase = number | string | boolean | IEquatable;
```
Comment (jakhog) [inline Source/concepts/Concept.ts]: Just for completeness, I guess we could allow bigint as well? 
```suggestion
type ConceptBase = number | bigint | string | boolean | IEquatable;
```

File: Source/concepts/Concept.ts [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+     */
+    equals(other: Concept<T> | T): boolean {
+        if (other == null) return false;
+        const comparableValue = Concept.isConcept(other) ? other.value : other as T;
+        if (comparableValue == null) return false;
+        if (typeof comparableValue === 'string' || typeof comparableValue === 'number' || typeof comparableValue === 'boolean') {
+            return comparableValue === this.value;
+        } else {
+            return (other as IEquatable).equals(this.value);
+        }
```
Comment (jakhog) [inline Source/concepts/Concept.ts]: Also with bigint, maybe nicer to do:
```suggestion
        switch (typeof comparableValue) {
            case 'number':
            case 'bigint':
            case 'string':
            case 'boolean':
                return comparableValue === this.value;
            default:
                return this.value.equals(comparableValue);
        }
```

File: Source/types/PrimitiveTypeMap.ts [hunk 3]
```diff
@@ -0,0 +1,8 @@
+// Copyright (c) Dolittle. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+export interface PrimitiveTypeMap {
+    string: string;
+    number: number;
+    boolean: boolean;
```
Comment (jakhog) [inline Source/types/PrimitiveTypeMap.ts:7]: Dont really know what this is, but should bigint be here as well?

**Review by jakhog — CHANGES_REQUESTED** [review]

File: Source/types/PrimitiveTypeMap.ts (hunk 3 above)
Comment (woksin) [reply Source/types/PrimitiveTypeMap.ts:7]: Yes. Doing it

File: Source/concepts/Concept.ts (hunk 2 above)
Comment (woksin) [reply Source/concepts/Concept.ts]: This did not work exactly as intended, broke the specs. This fixes it: https://github.com/dolittle-fundamentals/JavaScript.Fundamentals/pull/9/commits/be4275d0cb8ea8773d143eef3c8e808b20e42f8c

File: Source/types/PrimitiveTypeMap.ts (hunk 3 above)
Comment (woksin) [reply Source/types/PrimitiveTypeMap.ts:7]: Fixed by https://github.com/dolittle-fundamentals/JavaScript.Fundamentals/pull/9/commits/be4275d0cb8ea8773d143eef3c8e808b20e42f8c

**Review by jakhog — APPROVED** [review]

**Comment by jakhog** [comment]
Woops, tried to just approve the changes, but I actually have a few more questions...

**Review by jakhog — CHANGES_REQUESTED** [review]
I was wondering about the purpose of `IEquatableTo<>`, and why we don't just use `IEquatable` in stead?

I'm also thinking how this will look for someone not using TypeScript, as there is no runtime checking of these types. So it would be very possible to send in two different kinds of `ConceptAs<,>` both relying on the same base class, and at runtime it seems the current code would accept them as equal?

IMO we should just have `IEquatable` and implement `equals(other: any)` so that it checks they have the same type. In a sense, that `equals(..)` have the same semantics as `===`. If we start thinking about using `equals(...)` for non-primitive `xMap<,>` an similar, we could end up in a situation with strange types on the keys, couldn't we?

**Comment by woksin** [comment]
Yea I agree. It makes more sense to only have the IEquatable for non TS consumers. We should also agree on the semantics of the equals then. IMO Concept<T> should be able to be compared against Concept<T> and T. And any other type would return false

**Comment by jakhog** [comment]
Cool. I'm not entirely sure what the semantics of the equals should be, so we probably need to write that down so we agree, but we can do that later after we try it out a bit. As for `Concept<T,x>` being able to _equal_ both another `Concept<T,x>` and `T` both makes sense, and not, depending on what we mean by equality. Just to take one option, making `a.equals(b) === true` meaning `a` and `b` can be used interchangeably in the code, would not be true for a `a: Concept<T,>` and `b: T`.

We might consider allowing a `loose: boolean` argument to the `equals` method so the user can choose what kind of comparison to use, so that for `loose = false` (the default) it behaves like `===` and for `loose = true` it behaves like `==`. Which I think would mean something like `a.equals(b, false) === true` implies `a` and `b` can be used interchangeably, and `a.equals(b, true) === true` implies `b` can be converted to `a`.

Does that make sense?

**Comment by woksin** [comment]
But when you say Concept<T, x> do you mean the class Concept<T> or the type ConceptAs<T,U>?

**Comment by jakhog** [comment]
Hmm, that's a good point. As the `ConceptAs<T,U>` is just a type, we have put the `.equals` on `Concept<T>` right? Does that mean that we loose the _branding_ when we compare instances?

What would this code do?
```typescript
type SomeId = ConceptAs<Guid, 'SomeId'>;
type OtherId = ConceptAs<Guid, 'OtherId'>;

const id: Guid = Guid.create();
const a: SomeId = Concept.from(id);
const b: OtherId = Concept.from(id);

console.log(a.equals(b));
```
From what I understand now, it sounds like it would print `true`?

**Comment by woksin** [comment]
Under the hood a and b would be equal in the sense that they are both Concept<Guid> and they have the same value. They so not point to the same piece of memory though so === would be false. That is the best that we can do for the Concept.equals method. However we could provide a static strict equals on Concept that takes in two ConceptAs instances maybe? Or I don’t know if that works since we would need to know the brand of both of these

**Comment by woksin** [comment]
@jakhog Look at changes here https://github.com/dolittle-fundamentals/JavaScript.Fundamentals/pull/9/commits/98535169d0eb561d9e0d7ff66966d80189984e78 this turns the Concept<T, U> class into the Flavour kinda. Having the __uniqueConceptName: U as a protected member on the class ensures that the type checking works as intended, I've checked. By doing this we can also use the __uniqueConceptName to get the correct semantics for the .equals method.

Also making sure that U extends string to make sure that we cannot pass anything other than a string as U (but passing the string type itself works, but that would essentially be the same as `unknown`)

**Comment by woksin** [comment]
@einari Maybe this PR is interesting for you?

**Comment by woksin** [comment]
Can we pull this in now? @jakhog @einari

**Pull request merged**